File: src/where_clauses.rs
// To avoid "bloating" the function signatures, you can pull
// trait bounds out of the signature, and move them into a
// *where clause*:

use std::fmt::Display;

pub fn clone_and_print<T>(item: &T) -> T
where
    T: Clone + Display,
{
    let clone = item.clone();
    println!("🦀 Clone: {}", clone);
    clone
}

pub fn add<T>(a: T, b: T) -> T
where
    T: std::ops::Add<Output = T>,
{
    a + b
}

/// # Output
#[test]
fn test() {
    println!("🦀 {}", clone_and_print(&"foo"));
    println!("🦀 {}", add(1, 2));
}

// 🦀 Clone: foo
// 🦀 foo
// 🦀 3

File: src/where_clauses.ts
// To avoid "bloating" the function signatures, you can pull
// trait bounds out of the signature, and move them into a
// *where clause*:

interface ICloneAble<T> {
    clone: () => ICloneAble<T>;
}

export function clone_and_print<T>(item: ICloneAble<T>) {
    let clone = item.clone();
    console.log(`🐥 Clone:`, clone);
    return clone
}

export function add<T extends number>(a: T, b: T): number {
    return a + b
}

export class CloneAbleItem implements ICloneAble<string> {
    constructor(public value: string) { }
    clone() {
        return new CloneAbleItem('foo')
    }

    toString() {
        return this.value
    }
}

/// # Output
console.log(`🐥 ${clone_and_print(new CloneAbleItem('foo'))}`)
console.log(`🐥 ${add(1, 2)}`)

// 🐥 Clone: CloneAbleItem { value: 'foo' }
// 🐥 foo
// 🐥 3